rest/route/github_routes.go

```diff
@@ -0,0 +1,141 @@
+package route
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/patch"
+	"github.com/evergreen-ci/evergreen/rest"
+	"github.com/evergreen-ci/evergreen/rest/data"
+	"github.com/google/go-github/github"
+	"github.com/mongodb/grip"
+)
+
+type githubHookApi struct {
+	secret []byte
+
+	event interface{}
+	msgId string
+}
+
+func getGithubHooksRouteManager(route string, version int) *RouteManager {
+	methods := []MethodHandler{}
+	if secret := getWebhookSecret(); len(secret) > 0 {
+		methods = append(methods, MethodHandler{
+			Authenticator: &NoAuthAuthenticator{},
+			RequestHandler: &githubHookApi{
+				secret: secret,
+			},
+			MethodType: http.MethodPost,
+		})
+
+	} else {
+		grip.Warning("Github webhook secret is empty! Github webhooks have been disabled!")
+	}
+
+	return &RouteManager{
+		Route:   route,
+		Methods: methods,
+		Version: version,
+	}
+}
+
+func (gh *githubHookApi) Handler() RequestHandler {
+	return &githubHookApi{
+		secret: getWebhookSecret(),
+	}
+}
+
+func getWebhookSecret() []byte {
+	secret := ""
+	if settings := evergreen.GetEnvironment().Settings(); settings != nil {
+		secret = settings.Api.GithubWebhookSecret
+	}
+
+	return []byte(secret)
+}
+
+func (gh *githubHookApi) ParseAndValidate(ctx context.Context, r *http.Request) error {
+	eventType := r.Header.Get("X-Github-Event")
+	gh.msgId = r.Header.Get("X-Github-Delivery")
+
+	if len(gh.secret) == 0 {
+		return rest.APIError{
+			StatusCode: http.StatusInternalServerError,
+		}
+	}
+
+	body, err := github.ValidatePayload(r, gh.secret)
+	if err != nil {
+		grip.Errorf("Rejecting Github webhook POST: %+v", err)
+		return rest.APIError{
+			StatusCode: http.StatusBadRequest,
+			Message:    "failed to read request body",
+		}
+	}
+
+	gh.event, err = github.ParseWebHook(eventType, body)
+	if err != nil {
+		return rest.APIError{
+			StatusCode: http.StatusBadRequest,
+			Message:    err.Error(),
+		}
+	}
+
+	return nil
+}
+
+func (gh *githubHookApi) Execute(ctx context.Context, sc data.Connector) (ResponseData, error) {
+	switch event := gh.event.(type) {
+	case *github.PingEvent:
```

Comment:
Should we instead have a default case that handles everything that isn't a PullRequestEvent? It's conceivable that we will receive other things from GitHub too, and we want to ignore them all.

Reply:
What sort of handling are you thinking of? We return 200 OK to all requests which acknowledges the web hook to Github's servers. In the case of the Pull request or ping, we check the validity of the data since we're interested in it, and create errors only if it's invalid.

Reply:
Gotcha, sgtm.